File: scm-csharp/scheme/primitives/PrimitiveStringLessEqP.cs
namespace scheme;

public class PrimitiveStringLessEqP : Primitive
{
    public override string Name()
    {
        return "string<=?";
    }

    public override string Info()
    {
        return
            "Syntax: (string<=? s1 s2 ...)\n" +
            "Library: (scheme base)\n" +
            "Description: Returns #t if the strings are monotonically non-increasing in lexicographic order, otherwise returns #f.\n" +
            "Example:\n" +
            "  (string<=? \"a\" \"b\") => #t\n" +
            "  (string<=? \"abc\" \"abc\") => #t\n" +
            "  (string<=? \"b\" \"a\") => #f";
    }
    
    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 2, -1);
        object current = arguments[0];
        for (int i = 1; i < arguments.Length; i++)
        {
            object next = arguments[i];
            int cmp = String.Compare(new String(Value.AsString(current)), new String(Value.AsString(next)), StringComparison.Ordinal);
            if (!(cmp <= 0)) return false;
            current = next;
        }
        return true;
    }
}

File: scm-java/src/main/scheme/primitives/PrimitiveStringLessEqP.java
package scheme.primitives;

import scheme.*;

public class PrimitiveStringLessEqP extends Primitive {
    @Override
    public String name() {
        return "string<=?";
    }

    @Override
    public String info() {
        return "Syntax: (string<=? s1 s2 ...)\n" +
               "Library: (scheme base)\n" +
               "Description: Returns #t if the strings are monotonically non-increasing in lexicographic order, otherwise returns #f.\n" +
               "Example:\n" +
               "  (string<=? \"a\" \"b\") => #t\n" +
               "  (string<=? \"abc\" \"abc\") => #t\n" +
               "  (string<=? \"b\" \"a\") => #f";
    }
    
    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 2, -1);
        Object current = arguments[0];
        for (int i = 1; i < arguments.length; i++) {
            Object next = arguments[i];
            int cmp = new String(Value.asString(current)).compareTo(new String(Value.asString(next)));
            if (!(cmp <= 0)) return false;
            current = next;
        }
        return true;
    }
}
